### DISCUSSION TITLE
Integrating GEX and ATAC peaks : 10x multiome

### DISCUSSION THREAD

**Discussion by shregeno8** [GreenleafLab/ArchR#1605] [Questions / Documentation] [answered]
Hi
I am working on a 10X multiome(GEX and ATAC) data and I would really like to use ArchR. Is it possible to integrate the Gene Expression data to the ATAC seq generated from the same 10x multiome experiment?

Thanks!
She

**Comment by rcorces** [comment]
https://greenleaflab.github.io/ArchR_2020/Ex-Analyze-Multiome.html

**Reply by shregeno8** [reply]
Hello
Thanks.

I am receiving an error though when creating arrow files

> 2022-09-08 11:14:26 : createArrowFiles has encountered an error, checking if any ArrowFiles completed..
> ArchR logging successful to : ArchRLogs/ArchR-createArrows-4b422bb2eae-Date-2022-09-08_Time-11-09-32.log
> character(0)
> Warning message:
> In mclapply(..., mc.cores = threads, mc.preschedule = preschedule) :
>   8 function calls resulted in an error

Attaching the log now.

-  Also I tried the reformatFragmentFiles() and still didn't work

**Reply by shregeno8** [reply]
[ArchR-createArrows-4b422bb2eae-Date-2022-09-08_Time-11-09-32.log](https://github.com/GreenleafLab/ArchR/files/9525967/ArchR-createArrows-4b422bb2eae-Date-2022-09-08_Time-11-09-32.log)

**Reply by rcorces** [reply]
Sorry, I dont have enough information to help. Perhaps you can try setting `subthreading = FALSE`

**Reply by ZayaAyush** [reply]
I'm working on 10xMultiome data. This tutorial doesn't have enough information (https://greenleaflab.github.io/ArchR_2020/Ex-Analyze-Multiome.html). After making UMAP, is it able to make Peak calling, Motif enrichment, and trajectory analysis using 10xMultiome? 
I tried to use the Multiome  analysis, but I'm having difficulty to make LSI_Combined UMAP.
proj_AB <- addCombinedDims(proj_AB, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
Error in f(init, x[[i]]) : 
  number of rows of matrices must match (see arg 2)
Thanks!

**Reply by rcorces** [reply]
@ZayaAyush - you can see an in-progress draft of an updated multiome tutorial here:
https://github.com/GreenleafLab/ArchR_Website/blob/ui_updates/bookdown/XX_MultiomicDataAnalysis.Rmd

All normal "ATAC-only" functionality of ArchR is available in multiome projects.

**Reply by snghprtk** [reply]
@ZayaAyush 
I am getting the same error as yours with addCommbinedDims. Were you able to solve it?

proj <- addCombinedDims(proj, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
Filtering 1 dims correlated > 0.75 to log10(depth + 1)

Error in f(init, x[[i]]): number of rows of matrices must match (see arg 2)
Traceback:
1. addCombinedDims(proj, reducedDims = c("LSI_ATAC", "LSI_RNA"), 
 .     name = "LSI_Combined")
2. lapply(seq_along(reducedDims), function(x) {
 .     rD <- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims[x], 
 .         dimsToUse = dimsToUse, scaleDims = scaleDims, corCutOff = corCutOff)
 .     cV <- apply(as.matrix(rD), 2, var)
 .     normV <- 1/sqrt(sum(cV))
 .     rD * normV * dimWeights[x]
 . }) %>% Reduce("cbind", .)
3. Reduce("cbind", .)
4. f(init, x[[i]])

**Reply by doulijun777** [reply]
Hello, I also encountered this problem. I successfully do the ATAC and RNA diemension reduction process, and when I do the integrated analysis, it is reported that,

projMulti2 <- addCombinedDims(projMulti2, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
Error in f(init, x[[i]]) : 
  number of rows of matrices must match (see arg 2)

Do you have any way to solve this problem? 

Thank you so much.

**Reply by snghprtk** [reply]
@doulijun777 Sorry I haven't figured it out. Waiting for @rcorces and @ZayaAyush to comment.

**Reply by ZayaAyush** [reply, marked as answer]
sorry, I couldn't make progress with ArchR and switched to the Signac package and analyzed the multiome data well.

**Reply by cedmo001** [reply]
@snghprtk @doulijun777 Have you figured out a fix for this? I have the same issue!

**Reply by ZayaAyush** [reply]
@rcorces rcorces
@cedmo001 
@snghprtk 
Did you figure out this problem? I moved back to ArchR, i have still same issue!

projMulti3 <- addCombinedDims(projMulti3, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
Error in f(init, x[[i]]) : 
  number of rows of matrices must match (see arg 2)

Here is my scripts:
inputFiles <- getInputFiles("~/Documents/ArchRanalysisofscATAC")  
inputFiles <- c("~Documents/ArchRanalysisofscATAC/WT_atac_fragments.tsv.gz",
                "~Documents/ArchRanalysisofscATAC/cKO_atac_fragments.tsv.gz")
names(inputFiles) <- c("WT", "cKO")

ArrowFiles <- createArrowFiles(inputFiles, force = TRUE)
ArrowFiles <- c("WT.arrow","cKO.arrow")

projMulti2 <- ArchRProject(ArrowFiles = ArrowFiles, outputDirectory = "~/Documents/ArchRanalysisofscATAC")

#Import scRNA

input <- c("~/Documents/file1/filtered_feature_bc_matrix.h5",
            "~/Documents/file2/filtered_feature_bc_matrix.h5")

#set your names variable
names <- c("WT", "cKO")

#set your featureType variable
featureType <- "Gene Expression"

featureMats <- lapply(seq_along(input), function(y) {
  message("Importing Feature Matrix ", y, " of ", length(input))
  ArchR:::.importFM(featureMatrix = input[y], featureType = featureType, 
            name = names[y])
})

seRNAcombined<-cbind(assay(featureMats[[1]]), assay(featureMats[[2]]))

seRNA<-SummarizedExperiment(assays=list(counts=seRNAcombined), rowRanges= rowRanges(featureMats[[1]]))

projMulti2 <- addGeneExpressionMatrix(input = projMulti2, seRNA = seRNA)
```
projMulti2 <- addDoubletScores(projMulti2, force = TRUE)
projMulti3 <- filterDoublets(projMulti2)
	
## Analysis of multiome data in ArchR

projMulti3 <- addIterativeLSI(
  ArchRProj = projMulti3, 
  clusterParams = list(
    resolution = 0.6, 
    sampleCells = 10000,
    n.start = 10
  ),
  saveIterations = FALSE,
  useMatrix = "TileMatrix", 
  depthCol = "nFrags",
  name = "LSI_ATAC",force = TRUE
)

projMulti3 <- addIterativeLSI(
  ArchRProj = projMulti3, 
  clusterParams = list(
    resolution = 0.6, 
    sampleCells = 10000,
    n.start = 10
  ),
  saveIterations = FALSE,
  useMatrix = "GeneExpressionMatrix", 
  depthCol = "Gex_nUMI",
  varFeatures = 2500,
  firstSelection = "variable",
  binarize = FALSE,
  name = "LSI_RNA",force = TRUE
)
projMulti3 <- addCombinedDims(projMulti3, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
Error in f(init, x[[i]]) : 
  number of rows of matrices must match (see arg 2)

**Reply by dannyconrad** [reply]
@ZayaAyush I just ran into this problem as well and I think I have the solution - at least for my case. As the error message suggests, the problem was that the dimensions of the two SVD matrices you're trying to combine are not identical. You can check this like so:
`projMulti3@reducedDims$LSI_ATAC$matSVD %>% dim()`
`projMulti3@reducedDims$LSI_RNA$matSVD %>% dim()`

The number of rows should in theory be the same for both and should match the number of cells in the ArchR project. I think in my case, the problem was that there were cells with 0 or otherwise very few RNA counts - and these were getting removed at some point in the IterativeLSI process - likely because they would otherwise break the computation. 

So try first filtering your project to remove cells with <5 or <10 RNA counts, these won't be very informative to your analysis anyway.

**Reply by ZayaAyush** [reply]
Thank you @dannyconrad 
I checked the  SVD matrics, and they are different.
projMulti3@reducedDims$LSI_ATAC$matSVD %>% dim()
[1] 14749    30
projMulti3@reducedDims$LSI_RNA$matSVD %>% dim()
[1] 14733    30

Can you please provide instructions on how to obtain 14733 cells from both data sets to be filtered out?
Thank you for your advice,

**Reply by dannyconrad** [reply]
@ZayaAyush 
If you followed the example Multiome workflow that @rcorces linked previously, then you should you have a variable saved in your ArchR project called “Gex_nUMI”. Subset your project like so:
`projMulti4 <- subsetArchRProject(projMulti3, cells = projMulti3$cellNames[projMulti3$Gex_nUMI >= 10]`

Then re-run the code to produce the LSI_ATAC and LSI_RNA dim reductions in your new ArchR project. Hopefully this solves your problem and you can combine them successfully.

**Reply by ZayaAyush** [reply]
@dannyconrad Thank you! 
projMulti4 <- subsetArchRProject(projMulti3, cells = projMulti3$cellNames[projMulti3$Gex_nUMI >= 10])
Error: Error in rD[[x]][[1]][cellsKeep, , drop = FALSE] : 
  subscript out of bounds
There is still an error message.

**Reply by ZayaAyush** [reply]
Thank you All,
I filtered cells and added dimsToUse=1:30 in the addIterativeLSI step, which resolved the issue.

test3 <- proj_scATAC2[proj_scATAC2$TSSEnrichment > 6 & proj_scATAC2$nFrags > 2500 & proj_scATAC2$Gex_nUMI >= 7 & !is.na(proj_scATAC2$Gex_nUMI)]

test3 <- addIterativeLSI(
    ArchRProj = test3, 
    clusterParams = list(
      resolution = 0.2, 
      sampleCells = 10000,
      n.start = 10
    ),
    saveIterations = FALSE,
    useMatrix = "TileMatrix", 
    depthCol = "nFrags",
    name = "LSI_ATAC",, 
    varFeatures = 25000, 
    dimsToUse = 1:30
)
                        
#LSI-RNA
test3 <- addIterativeLSI(
    ArchRProj = test3, 
    clusterParams = list(
      resolution = 0.2, 
      sampleCells = 10000,
      n.start = 10
    ),
    saveIterations = FALSE,
    useMatrix = "GeneExpressionMatrix", 
    depthCol = "Gex_nUMI",
    varFeatures = 15000,
    firstSelection = "variable",
    binarize = FALSE,
    name = "LSI_RNA",dimsToUse = 1:30
)
test3@reducedDims$LSI_ATAC$matSVD %>% dim()
[1] 14708    30
> test3@reducedDims$LSI_RNA$matSVD %>% dim()
[1] 14708    30
test3 <- addCombinedDims(test3, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
